codalab/lib/bundle_util.py

```diff
@@ -37,6 +37,44 @@ def bundle_to_bundle_info(model, bundle):
     return info
 
 
+def get_ancestors_string_representation(client, bundle_uuid):
+    """
+    Traverses up the dependency graph and compiles a nested list of the bundle's ancestors
+    :param client: Instance of JsonApiClient
+    :param bundle_uuid: UUID of bundle
+    :return: String representation of nested list
+    :rtype: String
+    """
+    bundle_cache = {}
+    seen = {bundle_uuid}
+    result = []
+
+    def get_bundle(uuid):
+        # check the cache before querying for a bundle
+        if uuid not in bundle_cache:
+            bundle_cache[uuid] = client.fetch('bundles', uuid)
+        return bundle_cache[uuid]
+
+    def bundle_as_string(bundle, num_of_spaces):
+        # sample output:  - mnist(0x1ba223)
+        return ' ' * num_of_spaces + '- {}({})'.format(
+            bundle['metadata']['name'], bundle['uuid'][:8]
+        )
+
+    def recurse(uuid, h):
+        bundle = get_bundle(uuid)
+        # two spaces for each level up the dependency graph
+        result.append(bundle_as_string(bundle, h * 2))
+        for dependency in bundle['dependencies']:
+            parent_uuid = dependency['parent_uuid']
+            if parent_uuid not in seen:
```

Comment:
If the bundle is seen, you'd still want to print out the bundle but just not recurse.

Reply:
@percyliang: I added a **...** indicator to show that a branch has been seen before, so we're truncating it for the output.

Sample output:
```
- run-run-run-sor..eddec0-0x96c128(0x91ab0b)
  - run-run-run-run..304538-0xa317f9(0xd71aaa)
    - run-run-sort.py..2eda52-0xeddec0(0x304538)
      - run-run-sort.py..2eda52-0xac8fad(0xa49383)
      - run-lines.txt-0x1579e7(0x809f42)
        - run-date(0x72ad8c)
        - sort.py(0x2eda52)
  - run-run-sort.py..2eda52-0xeddec0(0x52a740)
    - run-run-run-run..304538-0xa317f9(0xd71aaa)
      ...
    - run-run-date-0x72ad8c(0x687dee)
```